### PR TITLE
new Test bot based on Controllers and ActivityHandler

### PR DESCRIPTION
**Pull request by johnataylor**

### CODE REVIEW

**Comment by cleemullins** [comment]
@johnataylor Build fails. Can you take a look?

File: tests/Microsoft.Bot.Builder.TestBot/Controllers/BotController.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    [Route("bot")]
+    [ApiController]
+    public class BotController : ControllerBase
+    {
+        private IBotFrameworkHttpAdapter _adapter;
+        private IBot _bot;
+
+        public BotController(IBotFrameworkHttpAdapter adapter, IBot bot)
+        {
+            _adapter = adapter;
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.TestBot/Controllers/BotController.cs]: Given that this is a test bot, weird stuff can happen here. The Null-check + throw will save us debugging time down the line... 
_adapter = adapter ?? throw new ...

File: tests/Microsoft.Bot.Builder.TestBot/Controllers/BotController.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Bot.Builder.Integration.AspNet.Core;
+
+namespace Microsoft.Bot.Builder.TestBot.Controllers
+{
+    [Route("bot")]
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.TestBot/Controllers/BotController.cs]: This is going to cause confusion. :) 

Devs (myself included) are used to the "/api/messages" path, and are used to putting that into the emulator. Can you add a comment here saying, "Here's the URL to put into the emulator for local debugging" and show the example.

File: tests/Microsoft.Bot.Builder.TestBot/Microsoft.Bot.Builder.TestBot.csproj [hunk 3]
```diff
@@ -1,21 +1,18 @@
-﻿<Project Sdk="Microsoft.NET.Sdk.Web">
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
   <PropertyGroup>
     <TargetFramework>netcoreapp2.1</TargetFramework>
-    <IsPackable>false</IsPackable>
-    <GenerateDocumentationFile>false</GenerateDocumentationFile>
+    <UserSecretsId>f277272c-9113-411f-addd-a4acfa9701a2</UserSecretsId>
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.TestBot/Microsoft.Bot.Builder.TestBot.csproj]: What's this? Seems scary and should be deleted. 

... if it's not scary and needs to be kept, it needs a comment explaining why it's not scary.

File: tests/Microsoft.Bot.Builder.TestBot/Microsoft.Bot.Builder.TestBot.csproj [hunk 4]
```diff
@@ -1,21 +1,18 @@
...
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.AspNetCore.All"/>
-  </ItemGroup>
-
-  <ItemGroup>
-    <DotNetCliToolReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Tools" Version="2.0.4" />
+    <PackageReference Include="Microsoft.AspNetCore.App" />
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.TestBot/Microsoft.Bot.Builder.TestBot.csproj]: Perfect! 
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/metapackage-app?view=aspnetcore-2.2

File: tests/Microsoft.Bot.Builder.TestBot/Program.cs [hunk 5]
```diff
@@ -1,21 +1,24 @@
-﻿// Copyright (c) Microsoft Corporation. All rights reserved.
-// Licensed under the MIT License.
-
+﻿using System;
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.TestBot/Program.cs]: // Copyright

File: tests/Microsoft.Bot.Builder.TestBot/Program.cs [hunk 6]
```diff
@@ -1,21 +1,24 @@
...
+using System.Collections.Generic;
+using System.IO;
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.TestBot/Program.cs]: Remove & Sort usings

File: tests/Microsoft.Bot.Builder.TestBot/Microsoft.Bot.Builder.TestBot.csproj [hunk 7]
```diff
@@ -1,21 +1,18 @@
...
+    <PackageReference Include="Microsoft.AspNetCore.Razor.Design" Version="2.1.2" PrivateAssets="All" />
+    <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="2.1.1" />
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.TestBot/Microsoft.Bot.Builder.TestBot.csproj]: Please include the Stylecop bits, as even though this is a "Test Bot" it's still code we own, publish, and need to hold to a high and consistent bar.

File: tests/Microsoft.Bot.Builder.TestBot/Startup.cs [hunk 8]
```diff
@@ -1,93 +1,39 @@
...
-            //    return new InteceptorAdapter(botFrameworkAdapter);
-            //});
+            services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_1);
 
-            IStorage dataStore = new MemoryStorage();
-            var conversationState = new ConversationState(dataStore);
+            // Load the credentials from configuration and create the credential provider.
+            var appId = Configuration["BotFramework:AppId"];
+            var password = Configuration["BotFramework:Password"];
+            var credentialProvider = new SimpleCredentialProvider(appId, password);
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.TestBot/Startup.cs]: This makes me laugh. 

Can we update the SipleCredentialProvider so this "glue" code is not necessary.

File: tests/Microsoft.Bot.Builder.TestBot/Startup.cs [hunk 9]
```diff
@@ -97,18 +43,16 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env)
             {
                 app.UseDeveloperExceptionPage();
             }
+            else
+            {
+                app.UseHsts();
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.TestBot/Startup.cs]: I'm a bit skeptical on this - it's an awesome feature, but I'm not at all clear on the implications. Do we need to gen/maintain a X.509 Cert?

File: tests/Microsoft.Bot.Builder.TestBot/Controllers/BotController.cs (hunk 2 above)
Comment (johnataylor) [reply tests/Microsoft.Bot.Builder.TestBot/Controllers/BotController.cs]: or I could change it to "api/messages" ???
- and yes it confused me when I deployed to Azure and it didn't work until i fixed up the URL .... aaahhh

File: tests/Microsoft.Bot.Builder.TestBot/Microsoft.Bot.Builder.TestBot.csproj (hunk 3 above)
Comment (johnataylor) [reply tests/Microsoft.Bot.Builder.TestBot/Microsoft.Bot.Builder.TestBot.csproj]: came from the asp project system... i can find out more

File: tests/Microsoft.Bot.Builder.TestBot/Startup.cs (hunk 8 above)
Comment (johnataylor) [reply tests/Microsoft.Bot.Builder.TestBot/Startup.cs]: you mean have the constructor take the dictionary as argument?

File: tests/Microsoft.Bot.Builder.TestBot/Startup.cs (hunk 9 above)
Comment (johnataylor) [reply tests/Microsoft.Bot.Builder.TestBot/Startup.cs]: this is basically the default template - I was deliberately doing as few edits as possible - maybe we should discuss what you want to do here

**Pull request merged**